From 744f3ee225ca31fcde8ae24c88545731dad07a49 Mon Sep 17 00:00:00 2001
From: Sven van Haastregt <sven.vanhaastregt@arm.com>
Date: Mon, 27 Jun 2022 17:07:31 +0100
Subject: [PATCH 061/172] [NFC] Factor out mutateArgsForImageOperands

(cherry picked from commit 2892a1ec1bab23a95dd0fe14eb8527c95ad5c2d6)
---
 lib/SPIRV/SPIRVToOCL.cpp | 38 +++++++++++++++++---------------------
 lib/SPIRV/SPIRVToOCL.h   |  3 +++
 2 files changed, 20 insertions(+), 21 deletions(-)

diff --git a/lib/SPIRV/SPIRVToOCL.cpp b/lib/SPIRV/SPIRVToOCL.cpp
index 33384d1f..642e1cdb 100644
--- a/lib/SPIRV/SPIRVToOCL.cpp
+++ b/lib/SPIRV/SPIRVToOCL.cpp
@@ -765,6 +765,19 @@ static char getTypeSuffix(Type *T) {
   return Suffix;
 }
 
+void SPIRVToOCLBase::mutateArgsForImageOperands(std::vector<Value *> &Args) {
+  if (Args.size() > 4) {
+    ConstantInt *ImOp = dyn_cast<ConstantInt>(Args[3]);
+    ConstantFP *LodVal = dyn_cast<ConstantFP>(Args[4]);
+    // Drop "Image Operands" argument.
+    Args.erase(Args.begin() + 3, Args.begin() + 4);
+    // If the image operand is LOD and its value is zero, drop it too.
+    if (ImOp && LodVal && LodVal->isNullValue() &&
+        ImOp->getZExtValue() == ImageOperandsMask::ImageOperandsLodMask)
+      Args.erase(Args.begin() + 3, Args.end());
+  }
+}
+
 // TODO: Handle unsigned integer return type. May need spec change.
 void SPIRVToOCLBase::visitCallSPIRVImageSampleExplicitLodBuiltIn(CallInst *CI,
                                                                  Op OC) {
@@ -787,16 +800,7 @@ void SPIRVToOCLBase::visitCallSPIRVImageSampleExplicitLodBuiltIn(CallInst *CI,
     auto Sampler = CallSampledImg->getArgOperand(1);
     Args[0] = Img;
     Args.insert(Args.begin() + 1, Sampler);
-    if (Args.size() > 4) {
-      ConstantInt *ImOp = dyn_cast<ConstantInt>(Args[3]);
-      ConstantFP *LodVal = dyn_cast<ConstantFP>(Args[4]);
-      // Drop "Image Operands" argument.
-      Args.erase(Args.begin() + 3, Args.begin() + 4);
-      // If the image operand is LOD and its value is zero, drop it too.
-      if (ImOp && LodVal && LodVal->isNullValue() &&
-          ImOp->getZExtValue() == ImageOperandsMask::ImageOperandsLodMask)
-        Args.erase(Args.begin() + 3, Args.end());
-    }
+    mutateArgsForImageOperands(Args);
     if (CallSampledImg->hasOneUse()) {
       CallSampledImg->replaceAllUsesWith(
           UndefValue::get(CallSampledImg->getType()));
@@ -831,17 +835,9 @@ void SPIRVToOCLBase::visitCallSPIRVImageWriteBuiltIn(CallInst *CI, Op OC) {
       M, CI,
       [=](CallInst *, std::vector<Value *> &Args) {
         llvm::Type *T = Args[2]->getType();
-        if (Args.size() > 4) {
-          ConstantInt *ImOp = dyn_cast<ConstantInt>(Args[3]);
-          ConstantFP *LodVal = dyn_cast<ConstantFP>(Args[4]);
-          // Drop "Image Operands" argument.
-          Args.erase(Args.begin() + 3, Args.begin() + 4);
-          // If the image operand is LOD and its value is zero, drop it too.
-          if (ImOp && LodVal && LodVal->isNullValue() &&
-              ImOp->getZExtValue() == ImageOperandsMask::ImageOperandsLodMask)
-            Args.erase(Args.begin() + 3, Args.end());
-          else
-            std::swap(Args[2], Args[3]);
+        mutateArgsForImageOperands(Args);
+        if (Args.size() > 3) {
+          std::swap(Args[2], Args[3]);
         }
         return std::string(kOCLBuiltinName::WriteImage) + getTypeSuffix(T);
       },
diff --git a/lib/SPIRV/SPIRVToOCL.h b/lib/SPIRV/SPIRVToOCL.h
index 275c5427..a1f02d6e 100644
--- a/lib/SPIRV/SPIRVToOCL.h
+++ b/lib/SPIRV/SPIRVToOCL.h
@@ -272,6 +272,9 @@ private:
 
   std::string translateOpaqueType(StringRef STName);
 
+  /// Mutate the argument list based on (optional) image operands.
+  void mutateArgsForImageOperands(std::vector<Value *> &Args);
+
 protected:
   Module *M;
   LLVMContext *Ctx;
-- 
2.20.1

